File: problems/basic/347/solution.ts
// Accepted: 68ms (70.15%), 56.28MB (6.30%)

function topKFrequent(nums: number[], k: number): number[] {
    const counts = new Map<number, number>();
    for (const num of nums) {
        counts.set(num, (counts.get(num) ?? 0)+1);
    }

    const countToNums = new Map<number, number[]>();
    for (const [num, count] of counts.entries()) {
        if (countToNums.has(count)) {
            countToNums.get(count)!.push(num);
        } else {
            countToNums.set(count, [num]);
        }
    }

    const tops: number[] = [];
    for (let count = nums.length; count > 0; --count) {
        if (!countToNums.has(count)) {
            continue;
        }

        for (const num of countToNums.get(count)!) {
            tops.push(num);
            k--;
        }

        if (k <= 0) {
            break;
        }
    }

    return tops;
};

File: problems/basic/347/solution.py
# Accepted: 99ms (38.98%), 27.16MB (5.22%)

class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        counts = defaultdict(int)
        for num in nums:
            counts[num] += 1

        countToNums = defaultdict(list)
        for num, count in counts.items():
            countToNums[count].append(num)

        tops = []
        for count in reversed(range(1, len(nums)+1)):
            if countToNums[count] == 0:
                continue

            for num in countToNums[count]:
                tops.append(num)
                k -= 1

            if k <= 0:
                break

        return tops
